[Model] tool calling support for ibm-granite/granite-20b-functioncalling

wseaton opened the pull request:
Add tool calling parser support for `ibm-granite/granite-20b-functioncalling`. 

Also adds an example chat template that is based off of the [granite fc paper](https://arxiv.org/abs/2407.00121).

wseaton committed:
remove original part of template

wseaton committed:
clean up debug logging

wseaton committed:
update docs; raise not implemented

wseaton committed:
initial commit

wseaton committed:
fix lints

wseaton committed:
sort imports

wseaton committed:
yapf fixes

wseaton committed:
another format change

wseaton commented on examples/tool_chat_template_granite.jinja:
```diff
@@ -0,0 +1,88 @@
+
+
+
+{% set sys_prompt = 'You are a helpful assistant with access to the following function calls. Your task is to produce a sequence of function calls necessary to generate response to the user utterance. Use the following function calls as required.' %}
```
Good callout! In my evals, I have mostly doing zero-shot single turn calls. 

Let me update the example template with the more general conversational one since that is probably what most people want.

wseaton committed:
update example prompt to be conversational instead of single turn

wseaton committed:
update docs for template; link paper

njhill commented:
@wseaton one of the tests is failing and looks most likely related: https://buildkite.com/vllm/fastcheck/builds/3985#0191dd9c-724a-4fbe-99d3-a1e1c94a2106

wseaton committed:
Merge remote-tracking branch 'upstream/main' into granite-fc

wseaton commented:
@maxdebayser are you still interested in contributing the streaming JSON parser for granite support? I have rebased off of main and might have some bandwidth to work on it, just let me know :)

wseaton committed:
add granite to test config

wseaton committed:
fixup json

maxdebayser commented:
Hi @wseaton , yes, I've been working on this. There is only one unit test that isn't passing. But I know what the cause is.

maxdebayser committed:
Add stream support for Granite 20b Tool Use

This commits builds on previous work by Will Eaton
and adds support for streaming. It also adds the
model to the tool use unit tests.

In this commit the tool parser is renamed from
simply granite to granite-20b-fc to differentiate
from other granite models.

Another minor change is that in the chat template
the function description using the function signature
is now optional.

Co-authored-by: Will Eaton <me@wseaton.com>

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser commented:
@wseaton, now all tests are passing. I've pushed a squashed merge of everything to this draft PR so that you can take a look and perhaps merge with yours: https://github.com/vllm-project/vllm/pull/8915

maxdebayser committed:
fix docs

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
more robust whispace handling

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

wseaton committed:
remove reference to defunct granite parser

wseaton committed:
remove old template

wseaton commented:
@maxdebayser I've cherry picked the commits from your PR, please review!

maxdebayser commented on tests/tool_use/utils.py:
```diff
@@ -73,6 +73,14 @@ def ensure_system_prompt(messages: List[Dict[str, Any]],
         "supports_parallel":
         False,
     },
+    "granite": {
+        "model":
+        "ibm-granite/granite-20b-functioncalling",
+        "arguments": [
+            "--tool-call-parser", "granite", "--chat-template",
+            str(VLLM_PATH / "examples/tool_chat_template_granite.jinja")
+        ],
+    },
```
This one here is a duplicate.
```suggestion
```

maxdebayser reviewed (approved):
Thanks a lot @wseaton . There is an extra entry on the unit test model dict, but other than that it looks good to me.

cc @njhill

wseaton committed:
Update tests/tool_use/utils.py to remove dupe

Co-authored-by: Maximilien de Bayser <maxdebayser@gmail.com>

wseaton commented:
@maxdebayser working on fixing merge conflicts w/ upstream now

wseaton committed:
Merge remote-tracking branch 'upstream/main' into granite-fc

wseaton committed:
fix double import

wseaton committed:
add completion request arg to abstract method

wseaton committed:
formatting fixes

wseaton committed:
import sorts

wseaton committed:
appease yapf

maxdebayser commented on vllm/entrypoints/openai/tool_parsers/granite_20b_fc_tool_parser.py:
```diff
@@ -0,0 +1,252 @@
+
+logger = init_logger(__name__)
+
+
```
With the recent changes in main this annotation must be added

```suggestion
@ToolParserManager.register_module("granite-20b-fc")
```

maxdebayser commented on vllm/entrypoints/openai/tool_parsers/granite_20b_fc_tool_parser.py:
```diff
@@ -0,0 +1,252 @@
+                                              ExtractedToolCallInformation,
+                                              FunctionCall, ToolCall)
+from vllm.entrypoints.openai.tool_parsers.abstract_tool_parser import (
+    ToolParser)
```
```suggestion
    ToolParser, ToolParserManager)
```

wseaton committed:
Apply suggestions from code review

Co-authored-by: Maximilien de Bayser <maxdebayser@gmail.com>

maxdebayser commented:
@K-Mistele , if you have the time, could you take a look at this PR?

njhill commented on vllm/entrypoints/openai/tool_parsers/utils.py:
```diff
@@ -85,3 +91,30 @@ def find_all_indices(string, substring):
+
+# partial_json_parser doesn't support extra data and
+# JSONDecorder.raw_decode doesn't support partial JSON
+def partial_json_loads(input_str, flags):
```
Could you add typing to these functions? (sorry I know they were just moved in this PR!)

njhill commented on vllm/entrypoints/openai/tool_parsers/utils.py:
```diff
@@ -85,3 +91,30 @@ def find_all_indices(string, substring):
+        if "Extra data" in e.msg:
+            dec = JSONDecoder()
+            return dec.raw_decode(input_str)
+        else:
```
nit: redundant else

njhill reviewed (approved):
Thanks alot @wseaton and @maxdebayser for the reviews

njhill commented:
@K-Mistele would be great if you could give this a quick look over and stamp :pray:

wseaton committed:
remove redudant indents; add type hints to utils

wseaton committed:
Merge branch 'granite-fc' of github.com:wseaton/vllm into granite-fc

wseaton committed:
formatting churn

wseaton committed:
Merge branch 'main' into granite-fc

wseaton committed:
change to old style type aliasing

wseaton committed:
Merge branch 'granite-fc' of github.com:wseaton/vllm into granite-fc

wseaton commented:
Getting an OOM error on the granite20b CI checks, looks like the model doesn't quite fit into the available memory of the instance type used for that CI.

njhill commented on docs/source/serving/openai_compatible_server.md:
```diff
@@ -238,6 +247,15 @@ AI21's Jamba-1.5 models are supported.
+* `ibm-granite/granite-20b-functioncalling`
+
+Flags: `--tool-call-parser granite-20b-fc`
+`examples/tool_chat_template_granite_20b_fc.jinja`: this is a modified chat template from the original on Huggingface, which is not vLLM compatible. It blends function description elements from the Hermes template and follows the same system prompt as "Response Generation" mode from [the paper](https://arxiv.org/abs/2407.00121). Parallel function calls are supported.
```
@wseaton should `--chat-template` be added here? (like other examples above)

wseaton replied:
Yes, looks like I dropped it via merge, sorry about that

njhill commented:
> Getting an OOM error on the granite20b CI checks, looks like the model doesn't quite fit into the available memory of the instance type used for that CI.

@wseaton we may need to comment this test out for now, as long as it's been verified manually on a larger GPU.

maxdebayser commented:
The test could also be run on a larger GPU, although that's probably a bit wasteful
```
$ git diff .
diff --git a/.buildkite/test-pipeline.yaml b/.buildkite/test-pipeline.yaml
index 8c98aa36a..8a396fada 100644
--- a/.buildkite/test-pipeline.yaml
+++ b/.buildkite/test-pipeline.yaml
@@ -293,6 +293,7 @@ steps:
 - label: OpenAI-Compatible Tool Use # 20 min
   fast_check: false
   mirror_hardwares: [ amd ]
+  gpu: a100
   source_file_dependencies:
     - vllm/
     - tests/tool_use
```

wseaton committed:
Doc reformat, add back missing line

wseaton committed:
Temporarily disable the granite20b-fc test task

wseaton commented:
@njhill Happy to rebuild my fork of vLLM and re-verify the PR manually on an A100, disabled the test for now

maxdebayser commented:
@wseaton I've verified a couple of hours ago. The tool_use tests run successfully on an A100. I have a quantized version of the model that passes the tests and I'm testing it if it can fit in the CI environment.

njhill commented:
> The test could also be run on a larger GPU, although that's probably a bit wasteful
> 

@maxdebayser yes that's what I thought too, since none of the other model sub-tests require it

njhill commented:
Thanks @wseaton! Could you try merging in the latest main branch again? That should hopefully help with the failing CI tests.

K-Mistele commented:
@wseaton just a minor point of clarification, this PR is for Granite 20B functioncalling -- is this the same function calling format as the granite 3.0 models? I saw that these were released recently:

- https://huggingface.co/collections/ibm-granite/granite-30-models-66fdb59bbb54785c3512114f
- https://huggingface.co/ibm-granite/granite-3.0-8b-instruct

I'm not as familiar with IBM/granite as with some other families of models. if they are compatible, then that's great and we should indicate that in the docs, as well as name the tool parser `granite` most likely. If not (and this is what I suspect is the case, based on a brief reading of their docs & github code for agentic stuff), then it might be worth adding a call-out in the docs that this is NOT compatible with granite 3.0

maxdebayser commented:
@K-Mistele , they are not compatible. We have another [PR](https://github.com/vllm-project/vllm/pull/9027) for those models but we're waiting for this one to be merged first because there are some code dependencies.

wseaton committed:
Merge branch 'vllm-project:main' into granite-fc

maxdebayser commented:
@njhill , @wseaton my tests with a quantized version of the model are now [passing](https://github.com/vllm-project/vllm/pull/9700). With quantization the model size was reduced to ~20GB: `mbayser/granite-20b-functioncalling-FP8-KV`, but I still had to add a bunch of flags to reduce memory usage:
`"--max_num_seqs", "1", "--max-model-len", "1024", "--enforce-eager", "--cpu-offload-gb", "20"`

wseaton commented:
Because of the nature of this branch and how long it's been running (with many merge commits from  main), I don't feel comfortable doing a rebase to fix the DCO issue. Can it be bypassed to get this merged?

K-Mistele commented:
> @K-Mistele , they are not compatible. We have another [PR](https://github.com/vllm-project/vllm/pull/9027) for those models but we're waiting for this one to be merged first because there are some code dependencies.

sounds good! Let me know if you'd like for me to take a look at it  whenever it's ready :)

njhill reviewed (approved):
Thanks @wseaton @maxdebayser for all of the work on this and thanks @K-Mistele for also reviewing.

njhill commented:
@wseaton I'll merge this to unblock the other granite PR, could consider re-enabling the test with @maxdebayser's quantized model as a follow-on update...
